bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/util/ComponentUtils.java

```diff
@@ -166,4 +177,103 @@ public static ContentPolicy getPolicy(String contentPath, ResourceResolver resou
+     *
+     * @param param of any type T
+     * @return array of size 1 having the same variable
+     */
+    public static <T> T[] convertToArray(T param) {
+        if (param.getClass().isArray()) {
+            return (T[]) param;
+        } else {
+            T[] array = (T[]) Array.newInstance(param.getClass(), 1);
+            array[0] = (T) param;
+            return array;
+        }
+    }
+
+    /**
+     * Fetches all the custom property group names defined in the template policy
+     *
+     * @param policy reference to the template policy {@link ContentPolicy}
+     * @param resourceResolver {@link ResourceResolver}
+     * @return {@code List<String>} a list of all the group names of custom properties defined in the policy
+     */
+    public static List<String> getCustomPropertyGroupsFromPolicy(ContentPolicy policy, ResourceResolver resourceResolver) {
+        if (policy == null) {
+            return new ArrayList<>();
+        }
+        Resource policyResource = resourceResolver.resolve(policy.getPath());
+        List<String> groupNames = new ArrayList<>();
+        List<Resource> customPropertiesGroupsList;
+
+        Resource customPropertiesResource = policyResource.getChild(HEADLESS_CUSTOM_PROPERTY_NODE_NAME);
+        if (customPropertiesResource != null) {
+            customPropertiesGroupsList = StreamSupport.stream(customPropertiesResource.getChildren()
+                .spliterator(), false).collect(Collectors.toList());
+            customPropertiesGroupsList.forEach((customPropertiesGroup) -> {
+                for (Map.Entry<String, Object> entry : customPropertiesGroup.getValueMap().entrySet()) {
+                    if (entry.getKey().equals(CUSTOM_PROPERTY_GROUP_NAME)) {
+                        groupNames.add(entry.getValue().toString());
+                    }
+                }
+            });
+        } else {
+            return new ArrayList<>();
+        }
+        return groupNames;
+    }
+
+    /**
+     * Fetches the key value pairs associated with the group names passed in
+     *
+     * @param groupNames {@link Set<String>}
+     * @param policyResource {@link Resource}
+     * @return {@code Map<String, String>} key value pairs associated with the given group names
+     */
+    public static Map<String, String> getCustomPropertyPairsFromPolicy(Set<String> groupNames, Resource policyResource) {
+        Map<String, String> keyValuePairs = new HashMap<>();
+        List<Resource> customPropertiesGroupsList;
+        Resource customPropertiesResource = policyResource.getChild(HEADLESS_CUSTOM_PROPERTY_NODE_NAME);
+        if (customPropertiesResource != null) {
+            customPropertiesGroupsList = StreamSupport.stream(customPropertiesResource.getChildren()
+                .spliterator(), false)
+                .collect(Collectors.toList());
+            customPropertiesGroupsList.forEach((customPropertiesGroup) -> {
```

Comment:
this contains a considerable duplicate code to read policy, iterate group list and then iterate map. pls consider generalisation where we pass only the function to be applied to each iteration.